### DISCUSSION TITLE
Frame loss / stuttering on Windows

### DISCUSSION THREAD

**Discussion by congoblue** [DistroAV/DistroAV#848] [General]
It's great to see some progress being made with this very useful plugin. And thanks for the audio fix for NDI5.

I am trying to build a Windows system with a couple of NDI cameras into OBS and really struggling to get smooth frame rates. Sometimes it works fine but other times every 3rd or 4th frame is dropped. I have setup a test system with just the cameras, the receiving computer and a network switch just to check nothing else is going on .

My windows machine is a reasonably capable i7 with Nvidia 1650 graphics and nothing is maxed out according to resource monitor. Plus, NDI5 Studio Monitor is able to do a smooth recording with no frame drops. I also have vMix and that can do a smooth recording. So it really looks like some issue in the way OBS handles the incoming NDI frames.

I was looking at the NDI5 SDK and they mention using this color format for most efficient frame handling
`recv_desc.color_format = NDIlib_recv_color_format_fastest;`

also I wonder about
`obs_source_set_async_unbuffered();`
which the plugin sets to True if the low-latency option is selected.

I have tried changing these settings using the very handy build system but neither makes much difference. So I am still digging for why OBS struggles with this. There are a few open source NDI-monitor apps on GitHub which all seem to use event-based code for receiving video rather than the polling approach that the OBS plugin uses.

**Comment by congoblue** [comment]
Sorry to reply to my own post, but I may have actually found a solution myself. I made a schoolboy error while testing and had OBS set to 50fps when my cameras were 30fps. Once on the same frame rate I have found that setting

`obs_source_set_async_unbuffered(s->source, false);`
on line 532 of obs_ndi_source.cpp (@paulpv "develop" branch) gives smooth video.

The current code sets this to False if the NDI source is set to "Safe" latency but sets True if the NDI source is set to "Low (Experimental)"
There may be reasons for this that I don't know but it seems to cause a lot of frame loss, at least on my Windows system.

**Comment by Trouffman** [comment]
Thanks for the feedback & sharing the details.
This is now managed by a settings in the plugin.
